src/uucore/src/lib/features/checksum.rs

```diff
@@ -422,14 +421,86 @@ pub fn detect_algo(algo: &str, length: Option<usize>) -> UResult<HashAlgorithm>
 //    algo must be uppercase or b (for blake2b)
 // 2. <checksum> [* ]<filename>
 // 3. <checksum> [*]<filename> (only one space)
-const ALGO_BASED_REGEX: &str = r"^\s*\\?(?P<algo>(?:[A-Z0-9]+|BLAKE2b))(?:-(?P<bits>\d+))?\s?\((?P<filename>(?-u:.*))\)\s*=\s*(?P<checksum>[a-fA-F0-9]+)$";
-const ALGO_BASED_REGEX_BASE64: &str = r"^\s*\\?(?P<algo>(?:[A-Z0-9]+|BLAKE2b))(?:-(?P<bits>\d+))?\s?\((?P<filename>(?-u:.*))\)\s*=\s*(?P<checksum>[A-Za-z0-9+/]+={0,2})$";
+const ALGO_BASED_REGEX: &str = r"^\s*\\?(?P<algo>(?:[A-Z0-9]+|BLAKE2b))(?:-(?P<bits>\d+))?\s?\((?P<filename>(?-u:.*))\)\s*=\s*(?P<checksum>[A-Za-z0-9+/]+={0,2})$";
 
 const DOUBLE_SPACE_REGEX: &str = r"^(?P<checksum>[a-fA-F0-9]+)\s{2}(?P<filename>(?-u:.*))$";
 
 // In this case, we ignore the *
 const SINGLE_SPACE_REGEX: &str = r"^(?P<checksum>[a-fA-F0-9]+)\s(?P<filename>\*?(?-u:.*))$";
 
+lazy_static! {
+    static ref R_ALGO_BASED: Regex = Regex::new(ALGO_BASED_REGEX).unwrap();
+    static ref R_DOUBLE_SPACE: Regex = Regex::new(DOUBLE_SPACE_REGEX).unwrap();
+    static ref R_SINGLE_SPACE: Regex = Regex::new(SINGLE_SPACE_REGEX).unwrap();
+}
+
+/// Hold the data extracted from a checksum line.
+struct LineInfo {
+    algo_name: Option<String>,
+    algo_bitlen: Option<usize>,
+    checksum: String,
+    filename: Vec<u8>,
+
+    regex: &'static Regex,
+}
+
+impl LineInfo {
+    fn parse(s: impl AsRef<OsStr>, cached_regex: &mut Option<&'static Regex>) -> Option<Self> {
```

Comment:
I'm not sure about the `cached_regex` handling. I think it's logic that should be implemented at the caller of this function. It's related to my previous idea, and hence probably something for a future refactoring.

Reply:
I can handle the setting of cached_regex in the caller of the function, but I'll still have to pass some information to the `parse` function to prevent it to choose a new format when the one being cached shall have the priority.

Or maybe I did not exactly understood your comment here. I will leave it as so for the moment, until your idea is clearer to me

Reply:
Let me try again :)

I think on a conceptual level we have three concepts: `File`, `Line`, and `Rule`. `File` has many `Line`s and a `Rule`. `Rule` defines that there can't be a mix of "single space" and "double space" `Line`s in a `File`. To enforce the `Rule`, we have to know the `Line`s and it's `File` that knows the `Line`s.

Translated to the implementation level, this means we could have a "dumb" `parse` function that simply returns `LineInfo`s and it's then up to the caller to enforce the conceptual `Rule` by checking the `format` property of the `LineInfo`s.

I hope this makes sense, and of course, it's possible I miss something.

Reply:
It's not exactly what's happening, actually. I may have to rephrase the comment explaining this.

When given a line, `LineInfo::parse` will try all 3 regexes in a specific order : AlgoBased, then DoubleSpace, and finally SingleSpace.

Also, the set of strings matched by DoubleSpace is a subset of the one matched by SingleSpace.
What's expected is that if the first line only matches SingleSpace and the following ones match DoubleSpace,
we should nevertheless parse the following with SingleSpace (and consider the extra space as part of the filename). We can't just discard these lines after parsing just by looking at the LineFormat they used. 

I haven't been able to find proper reference of this in the man pages, I've based my understanding of the requirements on the testsuites of `cksum` and `hashsum`.

If I find a documented reference of the expected behavior, I will come back to this and rework it accordingly.

FYI, this behavior is tested by `test_hashsum::test_check_md5sum_only_one_space`

Reply:
Ah, ok, that makes sense. Thanks for the explanation!

Reply:
Thank you for asking :grin: